samcli/lib/test_runner/test_runner_template_generator.py

```diff
@@ -0,0 +1,296 @@
+        rf"^arn:{partition_regex}:sqs:{region_regex}:{account_regex}:[\w-]+$": {
+            "Resource": "AWS::SQS::Queue",
+            "Action": ["sqs:SendMessage"],
+        },
+        rf"^arn:{partition_regex}:s3:::[\w-]+$": {
+            "Resource": "AWS::S3::Bucket",
+            "Action": ["s3:PutObject", "s3:GetObject"],
+        },
+        rf"^arn:{partition_regex}:dynamodb:{region_regex}:{account_regex}:table\/[\w-]+$": {
+            "Resource": "AWS::Dynamodb::Table",
+            "Action": [
+                "dynamodb:GetItem",
+                "dynamodb:PutItem",
+            ],
+        },
+        rf"^arn:{partition_regex}:states:{region_regex}:{account_regex}:stateMachine:[\w-]+$": {
+            "Resource": "AWS::StepFunctions::StateMachine",
+            "Action": [
+                "stepfunction:StartExecution",
+                "stepfunction:StopExecution",
+            ],
+        },
+    }
+    return default_permissions_map
+
+
+def _get_action_mapping(resource_arn: str) -> Union[dict, None]:
+
+    """
+    Returns a dictionary mapping a Resource type to a list of IAM actions generated for that resource.
+
+    Returns `None` if the supplied ARN is for a Resource type without any IAM actions supported.
+
+    Parameters
+    ----------
+    resource_arn:
+        The arn of the resource for which the mapping is returned.
+
+    Returns
+    -------
+    dict:
+        A dictionary containing two keys: `'Resource'` and `'Action'`. `'Resource'` contains the AWS Resource for which the IAM actions list contained in `'Action'` maps to.
+
+        For example `{"Resource": "AWS::S3::Bucket", "Action": ["s3:PutObject", "s3:GetObject"]}`
+
+    """
+
+    permissions_map = _get_permissions_map()
+
+    for arn_regex, mapping in permissions_map.items():
+        if re.search(arn_regex, resource_arn) is not None:
+            LOG.info("Matched ARN `%s` to IAM actions %s", resource_arn, mapping["Action"])
+            return mapping
+
+    # If the supplied ARN is for a resource without any default actions supported
+    LOG.info("Found no match for ARN `%s` to any IAM actions.", resource_arn)
+    return None
+
+
+def _create_iam_statment_string(resource_arn: str) -> str:
+    """
+    Returns an IAM Statement in the form of a YAML string corresponding to the supplied resource ARN.
+
+    The generated Actions are commented out to establish a barrier of consent, preventing customers from accidentally granting permissions that they did not mean to.
+
+    The Action list is empty if there are no IAM Action permissions to generate for the given resource.
+
+    Parameters
+    ----------
+    resource_arn : str
+        The arn of the resource for which the IAM statement is generated.
+
+    Returns
+    -------
+    str:
+        An IAM Statement in the form of a YAML string.
+    """
+
+    # NOTE: The spacing within these strings is needed to keep the resulting YAML indentation valid
+    new_statement_template = (
```

Comment:
Regarding to the comment here, since this template is programatically generated, we might not need to use jinja here. You could keep starting template as yaml, read it as dictionary into the memory, then make changes to its fields (or add new fields), and then serialize back as yaml. In that way you don't need to worry about indentation.

Reply:
This is similar to an approach I had tried, but the main issue is that we need generated IAM actions to be commented out in YAML. There isn't a way to comment out specific lines programmatically when converting from dictionary to yaml. This meant I had to use some funky string manipulation hacks to place `#`s in front of the lines I wanted.

Although having necessary whitespace isn't ideal, I don't think it is likely to change. The base template can have more resources added, but that will not change the indentation for the existing ones. 

The advantage of using jinja is that we can just render it, inject our string, and done. There's no flipping between yaml to json and back to yaml and then do some string manipulation to comment out generated IAM actions